rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc

```diff
@@ -0,0 +1,275 @@
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+@bootstrapParser
+module lang::rascal::lsp::Rename
+
+import Exception;
+import IO;
+import List;
+import Location;
+import Node;
+import ParseTree;
+import Relation;
+import Set;
+import String;
+
+import lang::rascal::\syntax::Rascal;
+// import lang::rascalcore::check::Checker;
+import lang::rascalcore::check::Import;
+import analysis::typepal::TypePal;
+import analysis::typepal::TModel;
+
+import analysis::diff::edits::TextEdits;
+
+import vis::Text;
+
+import util::Maybe;
+import util::Reflective;
+
+// Only needed until we move to newer type checker
+// TODO Remove
+import ValueIO;
+
+data IllegalRenameReason
+    = InvalidName(str name)
+    | CausesDoubleDeclaration(rel[loc, loc] defs)
+    | CausesCapture(rel[loc use, loc oldDef, loc newDef] captures)
+    | DeclarationsNotEditable(set[loc] decls)
+    ;
+
+data RuntimeException
+    = IllegalRename(loc location, IllegalRenameReason reason)
+    | UnsupportedRename(set[tuple[loc location, str message]])
+    | UnexpectedFailure(str message)
+;
+
+private set[loc] getUses(TModel tm, loc def) {
+    return invert(getUseDef(tm))[def];
+}
+
+private set[Define] getDefines(TModel tm, loc useDef) {
+    set[loc] defLocs = tm.useDef[useDef] != {}
+                     ? tm.useDef[useDef] // `useDef` is a use location
+                     : {useDef};         // `useDef` is a definition location
+
+    return {tm.definitions[d] | d <- defLocs};
+}
+
+void throwIfNotEmpty(&T(&A) R, &A arg) {
+    if (arg != {}) {
+        throw R(arg);
+    }
+}
+
+void checkLegalName(str name) {
+    try {
+        parse(#Name, escapeName(name));
+    } catch ParseError(_): {
+        throw InvalidName(name);
+    }
+}
+
+void checkCausesDoubleDeclaration(set[Define] currentDefs, set[Define] newDefs) {
+    // Is newName already resolvable from a scope where <current-name> is currently declared?
+    throwIfNotEmpty(CausesDoubleDeclaration, {<nD.defined, cD> | Define nD <- newDefs
+                                                               , loc cD <- currentDefs.defined
+                                                               , isContainedIn(cD, nD.scope)
+                                             });
+}
+
+bool isImplicitDefinition(start[Module] _, rel[loc use, loc def] useDef, Define _: <_, _, _, variableId(), defined, _>) = defined in useDef.use; // use of name at implicit declaration loc
+
+bool isImplicitDefinition(start[Module] m, rel[loc, loc] _, Define _: <_, _, _, patternVariableId(), defined, _>) {
+    visit (m) {
```

Comment:
I get what is happening here, but it feels like there should be a more rascal way? especially the if part of it. like, why it's not part of the pattern match in `qualifiedName` or something?

Reply:
Fixed.